### PR TITLE
Ft sprint3 admin site

### PR DESCRIPTION
**Pull request by helloCaller**
Components + a little refactoring of navigation (no longer hide/show, but based on page "state")

### CODE REVIEW

File: admin-site/public/index.html [hunk 1]
```diff
@@ -9,6 +9,10 @@
 	<!-- Bootstrap CSS -->
 	<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.0/css/all.css" integrity="sha384-lZN37f5QGtY3VHgisS14W3ExzMWZxybE1SJSEsQp9S+oqd12jhcu+A56Ebc1zFSJ" crossorigin="anonymous">
 	<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css" integrity="sha384-GJzZqFGwb1QTTN6wy59ffF1BuGJpLSa9DkKMp0DgiMDm4iYMj70gZWKYbI706tWS" crossorigin="anonymous">
+	<!-- Bootstrap dependencies -->
```
Comment (jakemoves) [inline admin-site/public/index.html:12]: Does this mean modals are now working as described in the Bootstrap docs?

File: admin-site/src/App.svelte [hunk 2]
```diff
@@ -4,22 +4,24 @@
...
+  let storedEvents;
   let gotEvents = false;
+  //holds a store of events
+  $: events = [];
 
   // // TODO this needs to pickup an environment somehow (dev/staging/prod)
-  // // let serverURL = "https://staging.cohort.rocks/api/v2"
-  
-  // let serverURL = "http://localhost:3000/api/v2";
+  //serverURL populated with dropdown at login - default is staging.cohort.rocks
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: Can you check the window.location for 'localhost' or '.local'? If found, you could set the default serverURL to that local URL (i.e. 'http://localhost:3000' or 'http://some-computer-name.local')? That way we can still override using the dropdown but local dev is a bit easier.

File: admin-site/src/App.svelte [hunk 3]
```diff
@@ -189,38 +198,50 @@
...
 //to show/hide dev Tools.
   let devElement = false;
 
+//show/hide modal
+  // let modalhtml;
+  //double check that a delete Occasion has happened
+  let deleteOccasionHasHappened = false;
+
+  //hold DOM state
+  let pageState = 0;
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: I'm down for this approach for now, it makes it easier to understand the different states. Routing is the next step...

File: admin-site/src/App.svelte [hunk 4]
```diff
@@ -259,8 +279,13 @@
     } 
   };
 
+  function goBackAPage(){
+    pageState --;
+    // not sure this is the right space for this
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: Me neither, but if it works for now...

File: admin-site/src/App.svelte [hunk 5]
```diff
@@ -382,16 +418,16 @@
...
+    // verifying password logic 
+    var passwordCheck = document.getElementById('password').value;
+    if (passwordCheck == "5555"){
+      authenticated = true;
+      GetEvents();
+      pageState = 1;
     }
+  }
     
     function HideShowDev() {
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: camelCase for function names please :)

File: admin-site/src/App.svelte [hunk 6]
```diff
@@ -622,305 +497,80 @@ label{
...
-    <hr />
+<!-- #eventsList allows a list of events to be built and shown based on main events object -->
+{:else if pageState == 1}
+  <Page
+    pageID="eventsList"
+    headingText="Events">
 
     {#if events.length === 0}
-      <p>That's uneventful. Sorry, no events have been added yet</p>
+        <p>That's uneventful. Sorry, no events have been added yet</p>
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: I <3 your punny UI messages.

File: admin-site/src/App.svelte [hunk 7]
```diff
@@ -622,305 +497,80 @@ label{
...
-		      <span class="fa fa-angle-left" />
-          Back
-        </button>
-        <h3 class="text-center">Occasions</h3>
-      </div>
+  <Page
+    pageID = "occasionList" 
+    headingText="Occasions">
+
+    <div slot="backButton">
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: Nice use of slots!

File: admin-site/src/Button.svelte [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<script>
+export let buttonText = "";
+export let buttonType = 'btn-outline-primary btn-block';
```
Comment (jakemoves) [inline admin-site/src/Button.svelte]: Would this be better named as `buttonStyle`?

File: admin-site/src/Button.svelte [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+export let dataTarget = undefined;
+export let value = "";
+export let disabled = undefined;
+export let bsSizePosition = "col-md-12";
```
Comment (jakemoves) [inline admin-site/src/Button.svelte]: Unclear prop name...

File: admin-site/src/ButtonActions.svelte [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<script context="module">
```
Comment (jakemoves) [inline admin-site/src/ButtonActions.svelte]: So this executes in a global context?

File: admin-site/src/Login.svelte [hunk 11]
```diff
@@ -1,27 +1,27 @@
...
 
  <script>
-    export let authenticated = false;
+  import Page from './ParentPage.svelte';
+  import Button from './Button.svelte';
+  
+  export let authenticated = false;
+  export let serverURL;
 
     function verifyPassword(){
```
Comment (jakemoves) [inline admin-site/src/Login.svelte]: This seems like the logical place for this function... why is the same function (almost) also in App.svelte?

If you need `App` to do something when `authenticated = true`, my first thought would be to have `Login` emit an event that is consumed by `App`. https://svelte.dev/tutorial/component-events

File: admin-site/src/OpenOccasion.svelte [hunk 12]
```diff
@@ -0,0 +1,231 @@
+<script>
```
Comment (jakemoves) [inline admin-site/src/OpenOccasion.svelte]: Seems like this file should be renamed to `Occasion`, no?

File: admin-site/src/ParentPage.svelte [hunk 13]
```diff
@@ -0,0 +1,37 @@
+<script>
```
Comment (jakemoves) [inline admin-site/src/ParentPage.svelte:1]: Yes! This kind of templating is a good pattern I think.

File: admin-site/src/App.svelte [hunk 14]
```diff
@@ -29,9 +31,15 @@
...
-    console.log(serverURL)
+    
+    storedEvents = writable(grabbedFromServerEvents);
+    storedEvents.subscribe(value => {
+      events = value;
+      console.log(value);
+    });
+
     focusedEvent = events[0]
   }
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: When you start moving around logic, I think this cue broadcast function probably belongs in `Occasion`. The slider should emit an event (something like `activated`) that's consumed by `Occasion` and triggers cue broadcast.

Comment (jakemoves) [reply admin-site/src/App.svelte]: The pattern to keep in mind: Data Down, Actions Up. Parents provide data to children, children emit actions when Parents need to do something to update data. (Or a component handles everything internally)

Comment (jakemoves) [reply admin-site/src/App.svelte]: https://svelte.dev/tutorial/component-events

File: admin-site/src/Button.svelte [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+                type="button" 
+                class="btn {buttonType}" 
+                data-toggle="modal" 
+                data-target={dataTarget}
+                data-dismiss = {dataDismiss}
+                aria-label = {ariaLabel}
+                value = {value}
+                disabled = {disabled}>
+
+            <span class={iconLeft}/>
```
Comment (jakemoves) [inline admin-site/src/Button.svelte:46]: You may want to check out the Bootstrap Input Group component: https://getbootstrap.com/docs/4.4/components/input-group/

**Review by jakemoves — CHANGES_REQUESTED** [review]
Nice work pulling out components from `App`. I think your pageState solution is a good bridge towards routing. Next up (after these notes): pull functions and logic out of App and into your components, and have child components emit actions to tell their parents when to do a thing. Delete looks good.

File: admin-site/src/OpenOccasion.svelte [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
```
Comment (jakemoves) [inline admin-site/src/OpenOccasion.svelte]: See https://github.com/jakemoves/cohort-server/issues/176

File: admin-site/src/App.svelte [hunk 17]
```diff
@@ -270,8 +295,7 @@
       }).then( response => { 
         if(response.status == 200){
           response.json().then( details => {
-            document.getElementById("openOccasion").style.display = "none";
-            document.getElementById("closeOccasion").style.display = "block";
+            pageState=3;
```
Comment (jakemoves) [inline admin-site/src/App.svelte]: I think this function (and openOccasion, delete, etc) need to make sure the events store gets updated. Maybe create a function in `App` like `updateEvents` and use actions to trigger it?

File: admin-site/public/index.html (hunk 1 above)
Comment (helloCaller) [reply admin-site/public/index.html:12]: Yes. If that isn't the case, let me know

**Comment by jakemoves** [comment]
🙌🏻

File: admin-site/src/App.svelte (hunk 2 above)
Comment (helloCaller) [reply admin-site/src/App.svelte]: ok yeah. I have something implemented now.

File: admin-site/src/App.svelte (hunk 5 above)
Comment (helloCaller) [reply admin-site/src/App.svelte]: fixed

File: admin-site/src/Button.svelte (hunk 8 above)
Comment (helloCaller) [reply admin-site/src/Button.svelte]: Yeah, sure. And any other feedback on naming conventions is more than welcome.

File: admin-site/src/Button.svelte (hunk 9 above)
Comment (helloCaller) [reply admin-site/src/Button.svelte]: the bsSizePosition?....yeah.. bootstrap Size and Position, not great. grid? gridLayout?

File: admin-site/src/ButtonActions.svelte (hunk 10 above)
Comment (helloCaller) [reply admin-site/src/ButtonActions.svelte]: It allows for that script to get called/imported to other components. Not implemented, in the current state, just trying things out. Likely won't have to do much of it (if any) with functions in their respective components.

File: admin-site/src/Login.svelte (hunk 11 above)
Comment (helloCaller) [reply admin-site/src/Login.svelte]: Login isn't implemented as a component yet. I started then stopped to check in. I think that's a good solution though. I'll comment here again once it's included.

**Comment by jakemoves** [comment]
Maybe gridStyle?

**Comment by jakemoves** [comment]
Ah gotcha. Btw I’ve been working on authentication this week, I should have the back end ready to go by the end of the weekend. Let me know when this component has stabilized and I’ll do the initial refactor for auth on the front end.

File: admin-site/src/Button.svelte (hunk 9 above)
Comment (jakemoves) [reply admin-site/src/Button.svelte]: bootstrapGridStyle?

File: admin-site/src/Button.svelte (hunk 15 above)
Comment (helloCaller) [reply admin-site/src/Button.svelte:46]: Is there a place where we want buttons inline with inputs? So far, the markup inside the button element is only for text inside a button.

File: admin-site/src/OpenOccasion.svelte (hunk 16 above)
Comment (helloCaller) [reply admin-site/src/OpenOccasion.svelte]: Now has a back button.

File: admin-site/src/App.svelte (hunk 17 above)
Comment (helloCaller) [reply admin-site/src/App.svelte]: Created a getEventsAndStore function that grabs from the server to update the store.

File: admin-site/src/Button.svelte (hunk 15 above)
Comment (jakemoves) [reply admin-site/src/Button.svelte:46]: Ah, got it, this is just i.e. the chevron for the 'back' button.

**Pull request merged**